community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanWriterTest.java

```diff
@@ -0,0 +1,217 @@
+        {
+            // WHEN
+            for ( int i = 0; i < NODE_COUNT * 3; i++ )
+            {
+                NodeLabelUpdate update = randomUpdate( expected );
+                writer.write( update );
+            }
+        }
+
+        // THEN
+        for ( int i = 0; i < LABEL_COUNT; i++ )
+        {
+            long[] expectedNodeIds = nodesWithLabel( expected, i );
+            long[] actualNodeIds = asArray( new LabelScanValueIterator( RANGE_SIZE, inserter.nodesFor( i ) ) );
+            assertArrayEquals( "For label " + i, expectedNodeIds, actualNodeIds );
+        }
+    }
+
+    @Test
+    public void shouldNotAcceptUnsortedLabels() throws Exception
+    {
+        // GIVEN
+        ControlledInserter inserter = new ControlledInserter();
+        boolean failed = false;
+        try ( NativeLabelScanWriter writer = new NativeLabelScanWriter( inserter, RANGE_SIZE, 1 ) )
+        {
+            // WHEN
+            writer.write( NodeLabelUpdate.labelChanges( 0, EMPTY_LONG_ARRAY, new long[] {2, 1} ) );
+            // we can't do the usual "fail( blabla )" here since the actual write will happen
+            // when closing this writer, i.e. in the curly bracket below.
+        }
+        catch ( IllegalArgumentException e )
+        {
+            // THEN
+            assertTrue( e.getMessage().contains( "unsorted" ) );
+            failed = true;
+        }
+
+        assertTrue( failed );
+    }
+
+    private NodeLabelUpdate randomUpdate( long[] expected )
+    {
+        int nodeId = random.nextInt( expected.length );
+        long labels = expected[nodeId];
+        long[] before = getLabels( labels );
+        int changeCount = random.nextInt( 4 ) + 1;
+        for ( int i = 0; i < changeCount; i++ )
+        {
+            labels = flipRandom( labels, LABEL_COUNT, random.random() );
+        }
+        expected[nodeId] = labels;
+        return NodeLabelUpdate.labelChanges( nodeId, before, getLabels( labels ) );
+    }
+
+    private static class ControlledInserter implements IndexWriter<LabelScanKey,LabelScanValue>
+    {
+        private final Map<Integer,Map<LabelScanKey,LabelScanValue>> data = new HashMap<>();
+
+        @Override
+        public void close() throws IOException
+        {   // Nothing to close
+        }
+
+        @Override
+        public void put( LabelScanKey key, LabelScanValue value ) throws IOException
+        {
+            merge( key, value, ValueMergers.overwrite() );
+        }
+
+        @Override
+        public void merge( LabelScanKey key, LabelScanValue value, ValueMerger<LabelScanValue> amender )
+                throws IOException
+        {
+            // Clone since these instances are reused between calls, internally in the writer
+            key = clone( key );
+            value = clone( value );
+
+            Map<LabelScanKey,LabelScanValue> forLabel = data.get( key.labelId );
```

Comment:
`Map<LabelScanKey,LabelScanValue> forLabel = data.computeIfAbsent( key.labelId, k -> new TreeMap<>( KEY_COMPARATOR ) );` can be used instead of null check